lib/pbench/test/unit/server/database/test_server_config_db.py

```diff
@@ -1,56 +1,309 @@
+            added: A list of object proxies we expect to have been added (but
+                not yet committed)
+            queries: A count of queries we expect to have been made
+            committed: A set of objects we expect to have been committed
+            rolledback: True if we expect rollback to have been called
+        """
+        session = self.session
+        assert session
+        assert len(session.queries) == queries
+        assert len(session.added) == len(added)
+        xlated = [
+            a
+            for a in session.added
+            for k in added
+            if a.key == k.key and a.value == k.value
+        ]
+        assert session.added == xlated
+        assert sorted(list(session.committed.values())) == sorted(committed)
+        assert session.rolledback == rolledback
+
+    @pytest.fixture(autouse=True, scope="function")
+    def fake_db(self, monkeypatch, server_config):
+        """
+        Fixture to mock a DB session for testing.
+
+        We patch the SQLAlchemy db_session to our fake session. We also store a
+        server configuration object on the Database.Base. (TODO: I was unable
+        to monkeypatch this due to the odd context of our base object
+        modifications during DB initialization; this is "unsatisfying" but
+        shouldn't be a problem)
+        """
+        self.session = FakeSession()
+        with monkeypatch.context() as m:
+            m.setattr(Database, "db_session", self.session)
+            Database.Base.config = server_config
+            yield
+
+    def test_bad_key(self):
         """Test server config parameter key validation"""
         with pytest.raises(ServerConfigBadKey) as exc:
             ServerConfig.create(key="not-a-key", value="no-no")
         assert str(exc.value) == "Configuration key 'not-a-key' is unknown"
 
-    def test_construct(self, db_session):
+    def test_construct(self):
         """Test server config parameter constructor"""
         config = ServerConfig(key="dataset-lifetime", value="2")
         assert config.key == "dataset-lifetime"
         assert config.value == "2"
         assert str(config) == "dataset-lifetime: '2'"
+        self.check_session()
 
-    def test_create(self, db_session):
+    def test_create(self):
         """Test server config parameter creation"""
         config = ServerConfig.create(key="dataset-lifetime", value="2")
         assert config.key == "dataset-lifetime"
         assert config.value == "2"
         assert str(config) == "dataset-lifetime: '2'"
+        self.check_session(committed=[FakeRow(id=1, key="dataset-lifetime", value="2")])
 
-    def test_construct_duplicate(self, db_session):
+    def test_construct_duplicate(self):
         """Test server config parameter constructor"""
         ServerConfig.create(key="dataset-lifetime", value=1)
+        self.session.raise_on_commit = IntegrityError(
+            statement="", params="", orig=FakeDBOrig("UNIQUE constraint")
+        )
```

Comment:
Unless this is obvious to everyone, we should probably have a comment pointing out that the setting of `raise_on_commit` would affect _both_ calls to `create()` so it has to be done **after** the first one completes.

In fact, it might be worth moving the assignment inside the `with` block so that it is better associated with the call that it affects.